sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusAggSketch.scala

```diff
@@ -0,0 +1,183 @@
+ * Estimation Algorithm
+ * https://docs.google.com/document/d/1gyjfMHy43U9OWBXxfaeG-3MjGzejW1dlpyMwEYAAWEI/view?fullscreen#
+ *
+ * @param child the HLL++ sketch to be aggregated and evaluated.
+ */
+// scalastyle:on
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns the estimated cardinality of the HyperLogLog++ sketch.""",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(approx_count_distinct_sketch(col1))
+        FROM VALUES (1), (1), (2), (2), (3) tab(col1);
+       3
+  """,
+  group = "agg_funcs",
+  since = "3.3.1")
+case class HyperLogLogPlusPlusAggSketch(
+    child: Expression,
+    var relativeSD: Double = 0.05,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends HyperLogLogPlusPlusTrait {
+
+  def this(child: Expression) = {
+    this(child = child, relativeSD = 0.05, mutableAggBufferOffset = 0, inputAggBufferOffset = 0)
+  }
+
+  def this(child: Expression, relativeSD: Expression) = {
+    this(
+      child = child,
+      relativeSD = HyperLogLogPlusPlus.validateDoubleLiteral(relativeSD),
+      mutableAggBufferOffset = 0,
+      inputAggBufferOffset = 0
+    )
+  }
+
+  override def prettyName: String = "approx_count_distinct_agg_sketch"
+
+  /**
+   * Allocate enough words to store all registers, plus one extra long
+   * allowing us to check buffer compatibility during merge.
+   */
+  override val aggBufferAttributes: Seq[AttributeReference] = {
+    Seq.tabulate(hllppHelper.numWords + 1) { i =>
+      AttributeReference(s"MS[$i]", LongType)()
+    }
+  }
+
+  // Note: although this simply copies aggBufferAttributes, this common code can not be placed
+  // in the superclass because that will lead to initialization ordering issues.
+  override val inputAggBufferAttributes: Seq[AttributeReference] =
+  aggBufferAttributes.map(_.newInstance())
+
+  /**
+   * This aggregate merges the HLL++ sketches that were written by a previous
+   * HyperLogLogPlusPlusSketch instance, which utilized a HLL++ helper that may have been
+   * configured differently than the one provided by our super class. Because we can't
+   * re-instantiate the super class's HLL++ helper, let's maintain the first instance that
+   * we deserialize, and ensure that our instance is able to fit within the words
+   * allocated in the aggregation buffer by the super class's HLL++ helper instance.
+   */
+  private var relativeSDUpdated: Boolean = false
+
+  /**
+   * Update the HLL++ buffer.
+   */
+  override def update(buffer: InternalRow, input: InternalRow): Unit = {
+    val v = child.eval(input)
+    if (v != null) {
+
+      val (newRelativeSD, newInternalRow) =
+        HyperLogLogPlusPlusSketch.deserializeSketch(v.asInstanceOf[Array[Byte]])
+
+      relativeSDUpdated match {
+        case true =>
+          if (newRelativeSD != relativeSD) {
+            throw new IllegalStateException("One or more of the HyperLogLogPlusPlusSketch " +
```

Comment:
should we put the values in the error message to help with debugging? Can we think of any other information that we could add in here to help? If this is illegal state, we don't have to worry about the user-friendliness so much and can just add debug information...

Reply:
Good call, will do.